processing/src/main/java/io/druid/segment/DimensionQueryHelper.java

```diff
@@ -0,0 +1,359 @@
+  public Comparator<byte[]> getGroupingKeyByteComparator();
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * Perform a relative read on a grouping key ByteBuffer to retrieve a single dimension value, and
+   * add the retrieved value to a GroupBy result map.
+   *
+   * An implementation may choose to not add anything to the result map
+   * (e.g., as the String implementation does for empty rows)
+   *
+   * @param theEvent Result map for the GroupBy query being served
+   * @param outputName The output name of this dimension for the GroupBy query being served, as specified in the DimensionSpec
+   * @param dimSelector Dimension value selector, used for value lookups if needed
+   * @param keyBuffer Grouping key, already positioned at this dimension's offset
+   */
+  public void readDimValueFromGroupingKey(
+      Map<String, Object> theEvent,
+      String outputName,
+      Object dimSelector,
+      ByteBuffer keyBuffer
+  );
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * Read the current row from a dimension value selector and add the row values to the grouping key.
+   *
+   * This is called by GroupByEngine's updateValues() function, which uses recursion to traverse the dimensions in the grouping set.
+   *
+   * Before adding a dimension value to the grouping key, this function should duplicate() the provided key buffer and
+   * add the value to the duplicate key.
+   *
+   * After adding a dimension value to the duplicate grouping key, an implementation of this function should call
+   * updateValuesFn on the new key to perform the recursion.
+   *
+   * For multi-value rows, this function should duplicate the original grouping key before adding each value, and
+   * call updateValuesFn on each new key.
+   *
+   * See StringDimensionQueryHelper for a reference implementation.
+   *
+   * @param dimSelector Dimension value selector
+   * @param key ByteBuffer for the grouping key
+   * @param updateValuesFn Function provided by GroupByEngine for updateValues() recursion
+   * @return Return the result of calling updateValuesFn on the updated grouping key
+   */
+  public List<ByteBuffer> addDimValuesToGroupingKey(
+      Object dimSelector,
+      ByteBuffer key,
+      Function<ByteBuffer, List<ByteBuffer>> updateValuesFn
+  );
+
+
+  /**
+   * Retrieve the current row from a dimension value selector.
+   *
+   * @param dimSelector Dimension value selector
+   * @return Current row
+   */
+  public Object getRowFromDimSelector(Object dimSelector);
+
+
+  /**
+   * Used by GroupByEngineV2.
+   *
+   * Read the first value within a row values object (IndexedInts, IndexedLongs, etc.) and add that value
+   * to the keyBuffer at keyBufferPosition, and return the size of the row values object.
+   *
+   * @param valuesObj row values object
+   * @param keyBuffer grouping key
+   * @param keyBufferPosition offset within grouping key
+   * @return size of the row values object
+   */
+  public int initializeGroupingKeyV2Dimension(
```

Comment:
removed the row size functionality from this, there was a row size method already